### PR TITLE
Introduce PointD<DIM>, BoxD<DIM> and SphereD<DIM>

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: src/details/ArborX_Box.hpp [hunk 1]
```diff
@@ -25,79 +25,85 @@ namespace ArborX
...
+  static int const dim = DIM;
+
+  KOKKOS_FUNCTION
+  constexpr BoxD()
+  {
+    for (int d = 0; d < DIM; ++d)
+    {
+      _min_corner[d] = KokkosExt::ArithmeticTraits::finite_max<float>::value;
+      _max_corner[d] = KokkosExt::ArithmeticTraits::finite_min<float>::value;
+    }
```
Comment (aprokop) [inline src/details/ArborX_Box.hpp:40]: This replaced
```c++
  Point _min_corner = {{KokkosExt::ArithmeticTraits::finite_max<float>::value,
                        KokkosExt::ArithmeticTraits::finite_max<float>::value,
                        KokkosExt::ArithmeticTraits::finite_max<float>::value}};
  Point _max_corner = {{KokkosExt::ArithmeticTraits::finite_min<float>::value,
                        KokkosExt::ArithmeticTraits::finite_min<float>::value,
                        KokkosExt::ArithmeticTraits::finite_min<float>::value}};
```
but I'm not sure if it's still done at the compile time.

File: src/details/ArborX_Point.hpp [hunk 2]
```diff
@@ -18,31 +18,40 @@
...
+  KOKKOS_INLINE_FUNCTION constexpr PointD(float x, float y)
+      : _data{{x, y}}
   {}
 
-  KOKKOS_INLINE_FUNCTION
-  constexpr Point(float x, float y, float z)
+  template <int D = DIM, std::enable_if_t<D == 3, int> = 0>
+  KOKKOS_INLINE_FUNCTION constexpr PointD(float x, float y, float z)
       : _data{{x, y, z}}
   {}
```
Comment (aprokop) [inline src/details/ArborX_Point.hpp:56]: While it's useful at the moment, maybe we should introduce a `float[DIM]` constructor.

File: src/details/ArborX_Box.hpp [hunk 3]
```diff
@@ -25,79 +25,85 @@ namespace ArborX
  * size 2x spatial dimension with a default constructor to initialize
  * properly an "empty" box.
  */
-struct Box
+template <int DIM = 3>
+struct BoxD
```
Comment (aprokop) [inline src/details/ArborX_Box.hpp:29]: Should probably put all `*D` into Experimental namespace or something.

File: src/details/ArborX_Point.hpp [hunk 4]
```diff
@@ -18,31 +18,40 @@
...
-  constexpr Point(Abomination data)
-      : Point(static_cast<float>(data.xyz[0]), static_cast<float>(data.xyz[1]),
-              static_cast<float>(data.xyz[2]))
+  KOKKOS_INLINE_FUNCTION constexpr PointD(Abomination data)
+  {
+    for (int d = 0; d < DIM; ++d)
+      _data.coords[d] = static_cast<float>(data.xyz[d]);
+  }
+
+  template <int D = DIM, std::enable_if_t<D == 2> = 0>
```
Comment (aprokop) [inline src/details/ArborX_Point.hpp:48]: FIXME
```suggestion
  template <int D = DIM, std::enable_if_t<D == 2, int> = 0>
```

Comment (masterleinad) [reply src/details/ArborX_Point.hpp:48]: I would just do
```
template <typename... Args, typename Enable = std::enable_if_t<sizeof...(Args)==dim>>
constexpr PointD(Args... args):    
  _data {static_cast<float>(args)...}
{}
```

File: src/details/ArborX_Point.hpp (hunk 2 above)
Comment (masterleinad) [reply src/details/ArborX_Point.hpp:56]: I would add it when we need it.

**Comment by JBludau** [comment]
I don't know about this but we could think about renaming BoxD into HyperRectangle and SphereD into HyperSphere?
Even though I like the shorter names you chose

**Comment by aprokop** [comment]
> I don't know about this but we could think about renaming BoxD into HyperRectangle and SphereD into HyperSphere?

Not a fan of these, too many letters. And still the question of `Point`. In general, this will also depend on whether we want to expose these to a user. If we put them into the experimental or details namespace, we will care less about naming. But realistically, at some point we'll need to make an backwards incompatible change of renaming them back to the current `Point`, `Box`, `Sphere`, I think.

**Comment by JBludau** [comment]
> > I don't know about this but we could think about renaming BoxD into HyperRectangle and SphereD into HyperSphere?
> 
> Not a fan of these, too many letters. And still the question of `Point`. In general, this will also depend on whether we want to expose these to a user. If we put them into the experimental or details namespace, we will care less about naming. But realistically, at some point we'll need to make an backwards incompatible change of renaming them back to the current `Point`, `Box`, `Sphere`, I think.

yeah, I agree. I was just thinking that if we want to rename this would be a good time.

File: src/details/ArborX_Box.hpp [hunk 5]
```diff
@@ -25,79 +25,85 @@ namespace ArborX
...
 {
-  KOKKOS_DEFAULTED_FUNCTION
-  constexpr Box() = default;
+  static int const dim = DIM;
```
Comment (dalg24) [inline src/details/ArborX_Box.hpp:31]: Not used.  We'll add it later if we actually need it.

File: src/details/ArborX_Box.hpp (hunk 3 above)
Comment (dalg24) [reply src/details/ArborX_Box.hpp:29]: I would try.  We might get surprises with ADL

File: src/details/ArborX_Box.hpp [hunk 6]
```diff
@@ -25,79 +25,85 @@ namespace ArborX
  * size 2x spatial dimension with a default constructor to initialize
  * properly an "empty" box.
  */
-struct Box
+template <int DIM = 3>
```
Comment (dalg24) [inline src/details/ArborX_Box.hpp:28]: Why the default template parameter?

Comment (aprokop) [reply src/details/ArborX_Box.hpp:28]: No good reason.

File: src/details/ArborX_Box.hpp (hunk 5 above)
Comment (aprokop) [reply src/details/ArborX_Box.hpp:31]: I was thinking about deducing the dimension of the data from the return type of the `AccessTraits`. But sure, can be added later.

**Review by dalg24 — COMMENTED** [review]
You do not discuss why you decided to change `Point`, `Box`, and `Sphere` rather than adding just a pure addition of new geometry types.  Did you consider other designs?

Also did you mean to ignore `KDOP`.

**Comment by aprokop** [comment]
> You do not discuss why you decided to change `Point`, `Box`, and `Sphere` rather than adding just a pure addition of new geometry types. Did you consider other designs?

I'm not sure I entirely understand the question. Adding new geometry types would require duplicating the algorithms file. In addition, unless you also duplicate all BVH, you would also be forced to choose between the two inside BVH. 

> Also did you mean to ignore `KDOP`.

Yes. It is of a less priority right now, and can be dimensionalized later.

**Comment by dalg24** [comment]
> Adding new geometry types would require duplicating the algorithms file.

Not necessarily.  For instance we could use some sort of traits to extract the dimension and access coordinates from the provided geometries (in which case you could mix and match geometries I suppose).

> In addition, unless you also duplicate all BVH, you would also be forced to choose between the two inside BVH.

I don't understand.  Isn't the bounding volume type a template parameter (I understand you can't directly set it from the `BVH` alias but that is a detail)?

**Comment by aprokop** [comment]
> Not necessarily. For instance we could use some sort of traits to extract the dimension and access coordinates from the provided geometries (in which case you could mix and match geometries I suppose).

OK. If, say, you have to point classes, one the original and one high dimensional, I suppose you could introduce `is_point<Geometry>` to figure out whether something is a point, and then write something like
```c++
template<typename Geometry, std::enable_if_t<is_point<Geometry>{}>* = nullptr>
float distance(Geometry const& point1, Geometry const& point2)
{
  constexpr int dim = Geometry::dim;
  for (int d = 0; d < dim; ++d)
    ...
}
```

> I don't understand. Isn't the bounding volume type a template parameter (I understand you can't directly set it from the `BVH` alias but that is a detail)?

I suppose you could if you provide additional template arguments to things like `calculateSceneBoundingBox`.

Overall, what are the benefits of having multiple independent types?

**Comment by masterleinad** [comment]
> Overall, I feel like it becomes overly complicated without a strong benefit. What would be your reason for doing something like that? I would prefer to go for an easier route, and introduce it later if necessary.

I agree. This feels like the less intrusive version.

**Comment by aprokop** [comment]
> I agree. This feels like the less intrusive version.

I actually now prefer #708 more :) It will allow adding new geometries easier in the future, and is more general.

**Comment by masterleinad** [comment]
> I actually now prefer #708 more :) It will allow adding new geometries easier in the future, and is more general.

...and introduces a bunch of unnecessary SFINAE boilerplate. If you want to add more geometries that are actually different, it makes more sense to create all those tags and type traits then.

**Comment by aprokop** [comment]
By the majority vote closing in favor of #708.

**Pull request closed without merging**